File: backend/honeypot_orchestrator/services/nbtnns.py
from __future__ import annotations

import asyncio
from typing import Any

from honeypot_orchestrator.services.base import BaseUDPHoneypotService


class NBTNSSProtocol(asyncio.DatagramProtocol):
    def __init__(self, service: NBTNSSHoneypot) -> None:
        self.service = service
        self.transport: asyncio.DatagramTransport | None = None

    def connection_made(self, transport: asyncio.DatagramTransport) -> None:
        self.transport = transport

    def datagram_received(self, data: bytes, addr: tuple[str, int]) -> None:
        src_ip, src_port = addr
        if len(data) < 12:
            return
        
        # Determine query type: Name Query (0x0020) or Node Status Query (0x0021)
        # Questions section starts at 12. NetBIOS name is always 1 + 32 + 1 = 34 bytes.
        # Query type starts at offset 46 (2 bytes).
        if len(data) < 48:
            return
        
        query_type = data[46:48]
        query_name = _parse_netbios_ns_name(data)
        if not query_name:
            return
        
        if query_type == b"\x00\x21":
            asyncio.create_task(
                self.service.log_event(
                    "netbios_ns_node_status_query",
                    src_ip=src_ip,
                    src_port=src_port,
                    query_name=query_name,
                    summary=f"NetBIOS NS Node Status query (NBSTAT) from {src_ip}",
                )
            )
        else:
            asyncio.create_task(
                self.service.log_event(
                    "netbios_ns_query",
                    src_ip=src_ip,
                    src_port=src_port,
                    query_name=query_name,
                    summary=f"NetBIOS NS query for {query_name}",
                )
            )
        
        response_ip = "127.0.0.1" if self.service.host in {"0.0.0.0", "::"} else self.service.host
        if query_type == b"\x00\x21":
            response = _build_netbios_ns_node_status_response(data)
        else:
            response = _build_netbios_ns_response(data, response_ip)
            
        if response and self.transport:
            self.transport.sendto(response, addr)


class NBTNSSHoneypot(BaseUDPHoneypotService):
    def create_protocol(self) -> asyncio.DatagramProtocol:
        return NBTNSSProtocol(self)


def _decode_netbios_ns_name(encoded: bytes) -> str:
    if len(encoded) < 32:
        return ""
    decoded = bytearray()
    for index in range(0, 32, 2):
        high = encoded[index] - 0x41
        low = encoded[index + 1] - 0x41
        decoded.append((high << 4) | low)
    return decoded.rstrip(b" ").decode("ascii", errors="replace")


def _parse_netbios_ns_name(data: bytes) -> str:
    if len(data) < 12 + 33:
        return ""
    name_len = data[12]
    if name_len != 32:
        return ""
    return _decode_netbios_ns_name(data[13:13+32])


def _build_netbios_ns_response(query_data: bytes, ip_address_str: str) -> bytes:
    if len(query_data) < 12:
        return b""
    transaction_id = query_data[0:2]
    name_sec = query_data[12:12+34]
    
    try:
        ip_bytes = bytes(int(part) for part in ip_address_str.split("."))
    except Exception:
        ip_bytes = b"\x7f\x00\x00\x01"
        
    header = b"".join(
        [
            transaction_id,
            b"\x85\x00",
            b"\x00\x00",
            b"\x00\x01",
            b"\x00\x00",
            b"\x00\x00",
        ]
    )
    
    answer = b"".join(
        [
            name_sec,
            b"\x00\x20",
            b"\x00\x01",
            (300000).to_bytes(4, "big"),
            (6).to_bytes(2, "big"),
            b"\x00\x00",
            ip_bytes,
        ]
    )
    
    return header + answer


def _build_netbios_ns_node_status_response(query_data: bytes) -> bytes:
    if len(query_data) < 12:
        return b""
    transaction_id = query_data[0:2]
    name_sec = query_data[12:12+34]
    
    header = b"".join(
        [
            transaction_id,
            b"\x84\x00",  # Flags: Response, Authoritative
            b"\x00\x00",  # Questions: 0
            b"\x00\x01",  # Answer RRs: 1
            b"\x00\x00",  # Authority RRs: 0
            b"\x00\x00",  # Additional RRs: 0
        ]
    )
    
    # 3 names: WIN-SRV2019 (0x00 Workstation), CORP (0x00 Workgroup), WIN-SRV2019 (0x20 Server)
    n1 = b"WIN-SRV2019".ljust(15) + b"\x00" + b"\x04\x00"
    n2 = b"CORP".ljust(15) + b"\x00" + b"\x84\x00"
    n3 = b"WIN-SRV2019".ljust(15) + b"\x20" + b"\x04\x00"
    names_payload = n1 + n2 + n3
    
    # Stats: MAC Address + 40 bytes of 0x00
    mac_addr = b"\x00\x15\x5d\xa1\xb2\xc3"
    stats_payload = mac_addr + b"\x00" * 40
    
    data_payload = b"\x03" + names_payload + stats_payload
    data_len = len(data_payload)
    
    answer = b"".join(
        [
            name_sec,
            b"\x00\x21",  # TYPE: NBSTAT
            b"\x00\x01",  # CLASS: IN
            b"\x00\x00\x00\x00",  # TTL: 0
            data_len.to_bytes(2, "big"),
            data_payload,
        ]
    )
    
    return header + answer

File: backend/internal/services/udp/nbtnns.go
package udp

import (
	"honeypot-orchestrator/backend/internal/profiles"
	"context"
	"fmt"
	"net"
	"strings"

	"honeypot-orchestrator/backend/internal/defense"
	"honeypot-orchestrator/backend/internal/services"
	"honeypot-orchestrator/backend/internal/logger"
)

type NBTNSSHoneypot struct {
	baseService *services.BaseUDPService
	host        string
}

func NewNBTNSSHoneypot(
	name string,
	host string,
	port int,
	el *logger.EventLogger,
	ds *defense.DefenseSystem,
) *NBTNSSHoneypot {
	n := &NBTNSSHoneypot{host: host}
	n.baseService = services.NewBaseUDPService(name, host, port, el, ds, n.handleDatagram)
	return n
}

func (n *NBTNSSHoneypot) Name() string {
	return n.baseService.Name()
}

func (n *NBTNSSHoneypot) Port() int {
	return n.baseService.Port()
}

func (n *NBTNSSHoneypot) IsRunning() bool {
	return n.baseService.IsRunning()
}

func (n *NBTNSSHoneypot) Start(ctx context.Context) error {
	return n.baseService.Start(ctx)
}

func (n *NBTNSSHoneypot) Stop() error {
	return n.baseService.Stop()
}

func (n *NBTNSSHoneypot) Proto() string {
	return n.baseService.Proto()
}

func (n *NBTNSSHoneypot) handleDatagram(ctx context.Context, data []byte, srcAddr *net.UDPAddr) ([]byte, error) {
	srcIP := srcAddr.IP.String()
	srcPort := srcAddr.Port

	if len(data) < 48 {
		return nil, nil
	}

	queryType := data[46:48]
	queryName := parseNetbiosNsName(data)
	if queryName == "" {
		return nil, nil
	}

	if queryType[0] == 0x00 && queryType[1] == 0x21 {
		n.baseService.LogEvent("netbios_ns_node_status_query", map[string]interface{}{
			"src_ip":     srcIP,
			"src_port":   srcPort,
			"query_name": queryName,
			"summary":    fmt.Sprintf("NetBIOS NS Node Status query (NBSTAT) from %s", srcIP),
		})
	} else {
		n.baseService.LogEvent("netbios_ns_query", map[string]interface{}{
			"src_ip":     srcIP,
			"src_port":   srcPort,
			"query_name": queryName,
			"summary":    fmt.Sprintf("NetBIOS NS query for %s", queryName),
		})
	}

	respIP := "127.0.0.1"
	if n.host != "0.0.0.0" && n.host != "::" && n.host != "" {
		respIP = n.host
	}

	var response []byte
	if queryType[0] == 0x00 && queryType[1] == 0x21 {
		response = buildNetbiosNsNodeStatusResponse(data)
	} else {
		response = buildNetbiosNsResponse(data, respIP)
	}

	return response, nil
}

func decodeNetbiosNsName(encoded []byte) string {
	if len(encoded) < 32 {
		return ""
	}
	var decoded []byte
	for i := 0; i < 32; i += 2 {
		high := int(encoded[i]) - 0x41
		low := int(encoded[i+1]) - 0x41
		if high < 0 || low < 0 {
			return "<unknown>"
		}
		decoded = append(decoded, byte((high<<4)|low))
	}
	s := string(decoded)
	return strings.TrimRight(s, " ")
}

func parseNetbiosNsName(data []byte) string {
	if len(data) < 12+33 {
		return ""
	}
	nameLen := data[12]
	if nameLen != 32 {
		return ""
	}
	return decodeNetbiosNsName(data[13 : 13+32])
}

func buildNetbiosNsResponse(queryData []byte, ipAddressStr string) []byte {
	if len(queryData) < 12 {
		return nil
	}
	txID := queryData[0:2]
	nameSec := queryData[12 : 12+34]

	ip := net.ParseIP(ipAddressStr)
	ipBytes := ip.To4()
	if ipBytes == nil {
		ipBytes = []byte{127, 0, 0, 1}
	}

	header := []byte{
		txID[0], txID[1],
		0x85, 0x00,
		0x00, 0x00,
		0x00, 0x01,
		0x00, 0x00,
		0x00, 0x00,
	}

	answer := make([]byte, len(nameSec)+12)
	copy(answer[0:len(nameSec)], nameSec)
	offset := len(nameSec)
	answer[offset] = 0x00
	answer[offset+1] = 0x20
	answer[offset+2] = 0x00
	answer[offset+3] = 0x01
	answer[offset+4] = 0x00
	answer[offset+5] = 0x04
	answer[offset+6] = 0x93
	answer[offset+7] = 0xe0
	answer[offset+8] = 0x00
	answer[offset+9] = 0x06
	answer[offset+10] = 0x00
	answer[offset+11] = 0x00
	answer = append(answer, ipBytes...)

	return append(header, answer...)
}

func buildNetbiosNsNodeStatusResponse(queryData []byte) []byte {
	if len(queryData) < 12 {
		return nil
	}
	txID := queryData[0:2]
	nameSec := queryData[12 : 12+34]

	header := []byte{
		txID[0], txID[1],
		0x84, 0x00,
		0x00, 0x00,
		0x00, 0x01,
		0x00, 0x00,
		0x00, 0x00,
	}

	n1 := padNetbiosName("WIN-SRV2019", 0x00, 0x04, 0x00)
	n2 := padNetbiosName("CORP", 0x00, 0x84, 0x00)
	n3 := padNetbiosName("WIN-SRV2019", 0x20, 0x04, 0x00)
	namesPayload := append(n1, n2...)
	namesPayload = append(namesPayload, n3...)

	macAddr := []byte{0x00, 0x15, 0x5d, 0xa1, 0xb2, 0xc3}
	statsPayload := append(macAddr, make([]byte, 40)...)

	dataPayload := append([]byte{0x03}, namesPayload...)
	dataPayload = append(dataPayload, statsPayload...)
	dataLen := len(dataPayload)

	answer := make([]byte, len(nameSec)+10)
	copy(answer[0:len(nameSec)], nameSec)
	offset := len(nameSec)
	answer[offset] = 0x00
	answer[offset+1] = 0x21
	answer[offset+2] = 0x00
	answer[offset+3] = 0x01
	answer[offset+4] = 0x00
	answer[offset+5] = 0x00
	answer[offset+6] = 0x00
	answer[offset+7] = 0x00
	answer[offset+8] = byte(dataLen >> 8)
	answer[offset+9] = byte(dataLen & 0xFF)
	answer = append(answer, dataPayload...)

	return append(header, answer...)
}

func padNetbiosName(name string, suffix byte, flagsHigh, flagsLow byte) []byte {
	buf := make([]byte, 18)
	copy(buf, []byte(strings.ToUpper(name)))
	for i := len(name); i < 15; i++ {
		buf[i] = 0x20
	}
	buf[15] = suffix
	buf[16] = flagsHigh
	buf[17] = flagsLow
	return buf
}

func init() {
	services.Registry["nbtnns"] = func(name, host string, port int, el *logger.EventLogger, ds *defense.DefenseSystem, prof *profiles.HoneypotProfile) services.HoneypotService {
		return NewNBTNSSHoneypot(name, host, port, el, ds)
	}
}

func (s *NBTNSSHoneypot) PortNameHost() string {
	return s.baseService.PortNameHost()
}
